packit_service/service/api/webhooks.py

```diff
@@ -206,28 +208,94 @@ def post(self):
+        http_url = project_data["git_http_url"]
+
+        project_token = ProjectTokenModel.get_project(
+            namespace=namespace, repo_name=repo_name, project_url=http_url
+        )
+
+        if not project_token:
+            token = jwt.encode(
+                {"namespace": namespace, "repo_name": repo_name},
+                config.gitlab_token_secret,
+                algorithm="HS256",
+            )
+
+            ProjectTokenModel.create(
+                token=token,
+                namespace=namespace,
+                repo_name=repo_name,
+                project_url=http_url,
+            )
+
+            project = config.get_project(url=http_url)
+            packit_user = project.service.user.get_username()
+
+            logger.info("Confidential issue created.")
+
+            project.create_issue(
+                title="Packit-Service Authentication",
+                body=f"To configure Packit-Service with `{repo_name}` you need to\n"
+                f"configure the webhook settings. Head to {http_url}/hooks and add\n"
+                f"the Secret Token `{token}` to authenticate requests coming to Packit.\n\n"
+                f"Packit needs rights to set commit status to merge requests, Please grant\n"
+                f"[{packit_user}](https://gitlab.com/{packit_user}) `admin` permissions\n"
+                f"on the {namespace}/{repo_name} project. You can add the rights by clicking\n"
+                f"[here]({http_url}/-/project_members).",
+                private=True,
+            )
+
+    def validate_token(self):
         """
         https://docs.gitlab.com/ee/user/project/integrations/webhooks.html#secret-token
         """
         if "X-Gitlab-Token" not in request.headers:
             if config.validate_webhooks:
                 msg = "X-Gitlab-Token not in request.headers"
-                logger.warning(msg)
-                raise ValidationFailed(msg)
+                logger.info(msg)
+                self.create_confidential_issue_with_token()
+                return
             else:
                 # don't validate signatures when testing locally
                 logger.debug("Ain't validating token.")
                 return
 
         token = request.headers["X-Gitlab-Token"]
 
-        # Find a better solution
-        if token not in config.gitlab_webhook_tokens:
-            raise ValidationFailed("Payload token validation failed.")
+        if token in config.gitlab_webhook_tokens:
+            logger.debug("Deprecation Warning: Old Gitlab tokens used.")
+            return
+
+        gitlab_token_secret = config.gitlab_token_secret
+        if not gitlab_token_secret:
+            msg = "'gitlab_token_secret' not specified in the config."
+            logger.error(msg)
+            raise ValidationFailed(msg)
 
-        logger.debug("Payload token is OK.")
+        msg_failed = "Payload token {} failed".format
```

Comment:
This looks handy but a bit confusing -- I would use a regular string in those two places so one does not need to search and think about that `msg_failed` function.

You can then use a meaningful sentence -- e.g. `Signature of the payload token is not valid.` and `Payload token does not match the project."

Reply:
Yes `Signature of the payload token is not valid.` this sounds better.

Reply:
Made them in different variables.